### DISCUSSION TITLE
How to validate when I add properties in the `transform` method?

### DISCUSSION THREAD

**Discussion by lancejpollard** [colinhacks/zod#3135] [Q&A] [answered]
II have a schema like this:

    import { z } from 'zod'

    const IO = z.object({
      input: z.object({
        file: z.object({
          path: z.string(),
        }),
      }),
      output: z.object({
        file: z.object({
          path: z.optional(z.string()),
          format: z.optional(z.string()),
        }),
      }),
    })

Every deep property on `output.file` is optional, yet one or the other is required, and one is _guaranteed_ to be set after the following `transform` call.

The `.transform` looks like this:

    .transform((v: Record<string, any>, ctx: RefinementCtx) => {
      if (!v.output.file.path) {
        if (!v.output.file.format) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            message: `Must specify a an output file path or format.`,
            path: ['output', 'file'],
          })
        } else {
          v.output.file.path = replaceFileExtension(
            v.input.file.path as string,
            `.${v.output.file.format}`,
          )
        }
      }

      return v
    })

How can I get the zod types to be more accurate in this case, is there anything else I can do?

If I set `output.file.path` to be required, then I can't do this, which is what I'd like to do:

```
parse({
  input: { file: { path: 'foo.bar' } },
  output: { file: { format: 'pdf' } },
})
// ZodError: [
//   {
//     "code": "invalid_type",
//     "expected": "string",
//     "received": "undefined",
//     "path": [
//       "output",
//       "file",
//       "path"
//     ],
//     "message": "Required"
//   }
// ]
```

Ideally the transform would run before the validation kicks in. Any thoughts about that?

So basically in my method which calls `Schema.parse(ioObj)`, I have to then afterward do a TypeScript assertion on the output.file.path, is that proper how it should be done anyways?

**Comment by JacobWeisenburger** [comment, marked as answer]
There seems to be a missing function in your example:
![image](https://github.com/colinhacks/zod/assets/31667350/10202cf3-fd90-4afe-ac40-66798d146f01)

Please send that and then I'll be able to better understand.
